### ISSUE TITLE
Occasional eventhub sending message error

### ISSUE THREAD

**Issue by izvictor** [Azure/azure-sdk-for-net#20255]
Hello,
We have noticed an occasional error during sending an eventhub message :
```
"InnerException":{"IsTransient":false,"Reason":1,"EventHubName":"eventhub-pi-tepqa","Message":"Unable to create the items needed to communicate with the Event Hubs service. (eventhub-pi-tepqa)","Data":{},"InnerException":{"ClassName":"System.ObjectDisposedException","Message":"Cannot access a disposed object.","Data":null,"InnerException":null,"HelpURL":null,"StackTraceString":"   at System.Runtime.ExceptionServices.ExceptionDispatchInfo.Throw()
   at Microsoft.Azure.Amqp.AsyncResult.End[TAsyncResult](IAsyncResult result)
   at Microsoft.Azure.Amqp.AmqpObject.OpenAsyncResult.End(IAsyncResult result)
   at Microsoft.Azure.Amqp.AmqpObject.EndOpen(IAsyncResult result)
   at Microsoft.Azure.Amqp.Transport.AmqpTransportInitiator.HandleTransportOpened(IAsyncResult result)
   at Microsoft.Azure.Amqp.Transport.AmqpTransportInitiator.OnTransportOpenCompete(IAsyncResult result)
--- End of stack trace from previous location where exception was thrown ---
   at System.Runtime.ExceptionServices.ExceptionDispatchInfo.Throw()
   at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification(Task task)
   at **Azure.Messaging.EventHubs.Amqp.AmqpConnectionScope.<CreateAndOpenConnectionAsync>d__66.MoveNext()**
--- End of stack trace from previous location where exception was thrown ---
   at System.Runtime.ExceptionServices.ExceptionDispatchInfo.Throw()
   at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification(Task task)
   at Microsoft.Azure.Amqp.FaultTolerantAmqpObject`1.<OnCreateAsync>d__6.MoveNext()
--- End of stack trace from previous location where exception was thrown ---
   at System.Runtime.ExceptionServices.ExceptionDispatchInfo.Throw()
   at Microsoft.Azure.Amqp.Singleton`1.<GetOrCreateAsync>d__13.MoveNext()
--- End of stack trace from previous location where exception was thrown ---
   at System.Runtime.ExceptionServices.ExceptionDispatchInfo.Throw()
   at Microsoft.Azure.Amqp.Singleton`1.<GetOrCreateAsync>d__13.MoveNext()
--- End of stack trace from previous location where exception was thrown ---
   at System.Runtime.ExceptionServices.ExceptionDispatchInfo.Throw()
   at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification(Task task)
   at Azure.Messaging.EventHubs.Amqp.AmqpConnectionScope.<OpenProducerLinkAsync>d__64.MoveNext()
--- End of stack trace from previous location where exception was thrown ---
   at System.Runtime.ExceptionServices.ExceptionDispatchInfo.Throw()
   at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification(Task task)
   at Azure.Messaging.EventHubs.Amqp.AmqpProducer.<CreateLinkAndEnsureProducerStateAsync>d__43.MoveNext()","RemoteStackTraceString":null,"RemoteStackIndex":0,"ExceptionMethod":"8
Throw
mscorlib, Version=4.0.0.0, Culture=neutral, PublicKeyToken=b77a5c561934e089
System.Runtime.ExceptionServices.ExceptionDispatchInfo
Void Throw()","HResult":-2146232798,"Source":"mscorlib","WatsonBuckets":null,"ObjectName":"sasl2715"},"StackTrace":"   at System.Runtime.ExceptionServices.ExceptionDispatchInfo.Throw()
   at Azure.Messaging.EventHubs.Amqp.AmqpProducer.<CreateLinkAndEnsureProducerStateAsync>d__43.MoveNext()
--- End of stack trace from previous location where exception was thrown ---
   at System.Runtime.ExceptionServices.ExceptionDispatchInfo.Throw()
   at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification(Task task)
   at Microsoft.Azure.Amqp.FaultTolerantAmqpObject`1.<OnCreateAsync>d__6.MoveNext()
--- End of stack trace from previous location where exception was thrown ---
   at System.Runtime.ExceptionServices.ExceptionDispatchInfo.Throw()
   at Microsoft.Azure.Amqp.Singleton`1.<GetOrCreateAsync>d__13.MoveNext()
--- End of stack trace from previous location where exception was thrown ---
   at System.Runtime.ExceptionServices.ExceptionDispatchInfo.Throw()
   at Microsoft.Azure.Amqp.Singleton`1.<GetOrCreateAsync>d__13.MoveNext()
--- End of stack trace from previous location where exception was thrown ---
   at System.Runtime.ExceptionServices.ExceptionDispatchInfo.Throw()
   at Azure.Messaging.EventHubs.Amqp.AmqpProducer.<SendAsync>d__42.MoveNext()
--- End of stack trace from previous location where exception was thrown ---
   at System.Runtime.ExceptionServices.ExceptionDispatchInfo.Throw()
   at Azure.Messaging.EventHubs.Amqp.AmqpProducer.<SendAsync>d__42.MoveNext()
--- End of stack trace from previous location where exception was thrown ---
   at System.Runtime.ExceptionServices.ExceptionDispatchInfo.Throw()
   at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification(Task task)
   at Azure.Messaging.EventHubs.Amqp.AmqpProducer.<SendAsync>d__37.MoveNext()
--- End of stack trace from previous location where exception was thrown ---
   at System.Runtime.ExceptionServices.ExceptionDispatchInfo.Throw()
   at Azure.Messaging.EventHubs.Producer.EventHubProducerClient.<SendInternalAsync>d__53.MoveNext()
--- End of stack trace from previous location where exception was thrown ---
   at System.Runtime.ExceptionServices.ExceptionDispatchInfo.Throw()
   at Azure.Messaging.EventHubs.Producer.EventHubProducerClient.<SendInternalAsync>d__53.MoveNext()
--- End of stack trace from previous location where exception was thrown ---
   at System.Runtime.ExceptionServices.ExceptionDispatchInfo.Throw()
   at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification(Task task)
   at Azure.Messaging.EventHubs.Producer.EventHubProducerClient.<SendAsync>d__44.MoveNext()
--- End of stack trace from previous location where exception was thrown ---
   at System.Runtime.ExceptionServices.ExceptionDispatchInfo.Throw()
   at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification(Task task)
   at **Azure.Messaging.EventHubs.Producer.EventHubProducerClient.<SendAsync>d__43.MoveNext()**
--- End of stack trace from previous location where exception was thrown ---
   at System.Runtime.ExceptionServices.ExceptionDispatchInfo.Throw()
   at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification(Task task)
   at PI2Cloud.Services.DataPipes.DataReceivers.PIDataReceiver.<OnNext>d__4.MoveNext()","HelpLink":null,"Source":"mscorlib","HResult":-2146233088},"StackTrace":null,"HelpLink":null,"Source":null,"HResult":-2146233088}|
```

he error disappears after a certain amount of time. Is this some kind of transient error ?
the event hub producer configuration are the following:
```
_eventHubClient = new EventHubProducerClient(target.ConnectionString, target.EventHubName, new EventHubProducerClientOptions()
                    {
                        RetryOptions = new EventHubsRetryOptions
                        {
                            Mode = EventHubsRetryMode.Exponential,
                            MaximumRetries = 10,
                            Delay = TimeSpan.FromSeconds(1),
                            MaximumDelay = TimeSpan.FromMinutes(1)
                        },
                        ConnectionOptions = new EventHubConnectionOptions
                        {
                            TransportType = EventHubsTransportType.AmqpWebSockets
                        }
                    });
```
Initially I thought there is an issue with **azure-amqp** module but as mentioned in the issue [178](https://github.com/Azure/azure-amqp/issues/178) it seems more to be an EH SDK problem and as mentioned by @xinchen10 the EH SDK is supposed to handle this error and convert to a transient exception type of the SDK.
Is this rigth ?
Thanks for help.
Best regards.

**Comment by jsquire** [comment]
Hi @izvictor.  Thank you for bringing this to our attention and we regret that you're experiencing difficulties.  To be transparent, I don't agree with the assessment from the AMQP issue; an `ObjectDisposedException` should be terminal.  If the AMQP library intends to communicate that it is safe to retry, it should be surfacing a different exception.

I'll open a thread to discuss that internally and see if we can come to consensus.  If not, then we'll look at trying to special-case handling for when we're opening links and translate to a exception that we recognize as safe to retry.

**Comment by izvictor** [comment]
Hi @jsquire ,
Thanks for you answer and taking into consideration our issue hope you will find a consensus on this.
Best regards.

**Comment by jsquire** [comment]
This has been addressed in #20440 and will be included in the next Event Hubs release.  The maintainers of the AMQP library have agreed to consider surfacing an exception more contextually relevant in a future release.  Once those changes are released, we'll revisit the need for the special case handling in the messaging libraries.  

I'm going to close this out; please feel free to reopen if you'd like to discuss further.

**Comment by izvictor** [comment]
Thanks looking forward for the next Event Hub release. 
Best regards.
